Add ability to specify custom form_post template

tpazderka opened the pull request:
- [x] Any changes relevant to users are recorded in the `CHANGELOG.md`.
- [ ] The documentation has been updated, if necessary.
---
Close #334

schlenk commented on src/oic/oic/provider.py:
```diff
@@ -234,7 +234,8 @@ def __init__(self, name, sdb, cdb, authn_broker, userinfo, authz,
                  verify_ssl=True, capabilities=None, schema=OpenIDSchema,
-                 jwks_uri='', jwks_name='', baseurl=None, client_cert=None):
+                 jwks_uri='', jwks_name='', baseurl=None, client_cert=None,
+                 form_post_template=FORM_POST):
```
A name that hints it is about the form_post response_mode would be better.

The provider already has two template related args (template_lookup/template). Why not reuse the template_lookup to lookup a custom form_post response template too?

tpazderka replied:
Good catch. I can probably reuse the `template` arg.

I am not sure how (and for what) is the `template_lookup` used as its default value would cause error.

tpazderka commented:
OK, I cannot find any documentation about the `template` arg structure.

I guess that the structure depends on the value of `template_lookup` and in order to reuse this, it would require some work to implement a default `template_lookup` method.

Would you prefer the `template` to store raw templates as strings or paths to a file with template?

With that I can implement a default `template_lookup` function and go from there.

schlenk commented:
As far as i can see, the template/template_lookup thing is mostly used for the utils/http_util.py:Response code to render a template. And it is used rather inconsistently.

So template is a dict of 'template_names' (those should be documented somewhere), and some lookup key used by the template_lookup. The mako lookup uses relative filenames as far as i can see.

So i think the values should not be simple strings, but could be rather trivial function objects that have a very basic render method that mostly boils down to a `format()`.

If you look at the examples (op2), there seems to be a name collision with your proposed template name (maybe its the same thing).
oidc_example\op2\server.py
534:        "template": {"form_post": "form_response.mako"},

tpazderka commented:
OK, then.

I will try to implement this with default template_lookup as mako with directory somewhere in the project.

I will look into documenting the changes.

tpazderka commented:
Reworked a little bit.

I decided to remove `template_lookup` and `template` arguments (they look unused?) and replaced them with a `template_renderer` which should be capable of replacing them entirely.

The `verify_logout` template needs to be finished but I am not sure yet what was its function as it is nowhere to be found :) @rohe could you please explain the intended functions to me?

Documented and hopefully fixed the examples.

schlenk commented:
Looks good.

Not sure how the connection is to 'utils/http_utils.py:Response' which takes kwargs parameters for template_lookup/mako_template/template or falls back to some defaults. I didn't see right away how it might get there from the Provider class. Probably some templates are set for rendering the login forms etc.

tpazderka commented:
I will look into that. But if we figure out it is relevant and should be refactored, I suppose another issue would be a good place to keep the changes small and manageable.

schlenk commented:
ah well, seems some tests still fail, so fix those first, but the issue seems trivial.

tpazderka commented:
Yeah a typo :) And I am still waiting for the `verify_logout` template.

tpazderka commented:
OK, I have added some basic `verify_logout` template as I want to get this merged. We can improve upon it later if there is need for that.

schlenk commented:
OK. Feel free to merge.

The template rendering could probably use some hardening against malicious inputs later, but as far as i can see the input isn't really user controlled, so should be good enough for now. One could also move the onbody handler into an external file to help with Content-Security-Policies.

tpazderka commented:
I had CSP in mind but I think we cannot make any assumptions about deployment so hardcoding an external file might be difficult. This at least allows to override the template completely (styling, CSP and such).
